### ISSUE TITLE
Components added after calling removeAllComponents become non-interactive

### ISSUE THREAD

**Issue by xazap** [mabe02/lanterna#360]
Hi, 

I ran into an issue where I am unable to get focus or keyboard navigation when adding buttons to a Panel after calling removeAllComponents on that same Panel. The code below reproduces this with Lanterna 3.0.0.

The use-case for this is where my (MVC) controller updates a view implemented with Lanterna. I remove all components from a panel and then initialize all components again with updated model values. 

Experimenting a bit, I found that calling setFocusedInteractable on the Window that contains the panel sets focus. This is very expensive workaround as it would require me to keep track of which interactable component had focus before calling removeAllComponents. The component having focus is usually some other panel that may even be in some other window.

I observed this:
When I add buttons to a panel the first time, the first button automatically gets focus. When you clear all components and add another set of buttons, none of them get focus. Even worse: interactable components on other panels also have no focus, resulting in a stuck GUI. Is this a bug?

```java
import com.googlecode.lanterna.*;
import com.googlecode.lanterna.graphics.*;
import com.googlecode.lanterna.gui2.*;
import com.googlecode.lanterna.screen.*;
import com.googlecode.lanterna.terminal.*;

public class Reproduction {
	
	public static void main(String[] args) {
		try {
			Terminal terminal = new DefaultTerminalFactory().createTerminal();
			Screen screen = new TerminalScreen(terminal);
			screen.startScreen();
			
			WindowManager windowManager = new DefaultWindowManager(TerminalSize.ONE);
			WindowPostRenderer renderer = new WindowPostRenderer() {
				@Override
				public void postRender(ThemedTextGraphics textGraphics, TextGUI textGUI, Window window) {
					System.out.println("postRender");
				}
			};

			MultiWindowTextGUI gui = new MultiWindowTextGUI(new SeparateTextGUIThread.Factory(), screen, windowManager, renderer, new EmptySpace(TextColor.ANSI.MAGENTA));
			((AsynchronousTextGUIThread) gui.getGUIThread()).start();
			
			Panel panel = new Panel(new GridLayout(2));
			Button button1 = new Button("Button 1");
			panel.addComponent(button1);
			
			Window window = new BasicWindow();
			window.setComponent(panel);
			gui.addWindow(window);
			
			Thread.sleep(3000); // wait a bit so we can see it losing focus...
			panel.removeAllComponents();
			Button button2 = new Button("Button 2");
			panel.addComponent(button2);

			// workaround:
			// window.setFocusedInteractable(button2);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}
```

**Comment by avl42** [comment]
My observation is that for an Interactable to get initially focused it must already have been added to the structure of (maybe nested) Panels before the main Panel is set as the Component for the Window. Later adding of further components will not automatically set focus.

I have rearranged and modified your code:
```
package com.googlecode.lanterna.issue;

import com.googlecode.lanterna.gui2.*;
import com.googlecode.lanterna.screen.*;
import com.googlecode.lanterna.terminal.*;

public class Issue360 {

    public static void main(String[] args) {
        try {
            Terminal terminal = new DefaultTerminalFactory().createTerminal();
            Screen screen = new TerminalScreen(terminal);
            screen.startScreen();
            MultiWindowTextGUI gui = new MultiWindowTextGUI(screen);
            Window window = new BasicWindow();
            
            Button button1 = new Button("Button 1");
            Panel panel = new Panel(new GridLayout(2));

            panel.addComponent(button1);
            window.setComponent(panel);
            gui.addWindow(window);

            gui.updateScreen(); // shows focused Button 1
            Thread.sleep(1000);

            panel.removeComponent(button1);

            gui.updateScreen(); // shows empty Window
            Thread.sleep(1000);

            Button button2 = new Button("Button 2");
            panel.addComponent(button2);

            gui.updateScreen(); // shows non-focused Button 2
            Thread.sleep(1000);

            // mainly for demonstration purpose:
            window.setComponent(null); // don't let a draw() happen while null ...
            window.setComponent(panel);
            
            // alternatively, you can ask the Panel for any focusable child components:
            //window.setFocusedInteractable(panel.nextFocus(null));
            // (that's what window.setComponent() does, when called with a  Container like a Panel.)
            
            gui.updateScreen(); // now shows focused Button 2
            Thread.sleep(1000);

            System.exit(0);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
```

**Comment by mabe02** [comment]
When you assign a component to the window, it will find the first interactable component and focus on it. But if the component is a panel like above, when you remove and add components to it, we can't get the window to automatically adjust the focus (not easily at least) because it doesn't know what you are doing with the composite component's content.

You could solve it by calling `window.setFocusedInteractable(panel.nextFocus(null));` which will (re-)focus the first interactable in the panel.

**Comment by xazap** [comment]
Thanks for pointing out `window.setFocusedInteractable(panel.nextFocus(null));`. I have now implemented this in my project. I still get a stuck GUI sometimes though. Digging some more I found that there are actually two causes for my frozen GUI: Not having focus on an interactable component is one cause, which is now solved. Another cause is that the call to `removeAllComponents();` doesn't always complete. Looking at the code for the Panel class, I see that it synchronizes internally on "components":

```java
    public Panel removeAllComponents() {
        synchronized(components) {
            for(Component component : new ArrayList<Component>(components)) {
                removeComponent(component);
            }
        }
        return this;
    }
```
When I call `removeAllComponents` from my Controller thread it often waits (forever?) on getting the lock for "components". By suspending all threads before the synchronized block, I have found that a thread named "LaternaGUI" own the locks on "components" and is apparently not releasing it. This happens at a 50% ratio so it's very likely a thread-contention issue. 

What puzzles me though, it seems to wait forever on release of the lock. Worst case I would expect  a stalled completion when the LanternaGUI does its thing with the components and then releases its lock allowing my Controller thread to do its thing.

Any ideas on this?

**Comment by avl42** [comment]
I find those locks a bit suspicious, too. Most likely you're sometimes
facing a dining philosophers problem some thread holds one lock and tries
to aquire another lock, and a second thread holds the second lock and tries
to aquire the first one.    java can give some information if you can send
it signal 3 (on unix) or maybe with some tools in $JDK_HOME/bin/...   it
would be helpful if you could find out which thread is holding which locks
(aka monitors) and waiting for which other monitor.  if these threads are
within lanterna code, then a stackdump would also help...

xazap <notifications@github.com> schrieb am So., 20. Mai 2018, 17:43:

**Comment by avl42** [comment]
Well, that's bad.  I don't yet see an easy fix, but some approaches maybe towards a fix (all targetting lanterna, not your user code):

 * don't synchronize on the whole iterations of components, but rather only on making a local shallow copy of the collection and then iterate the copy. At least if the whole loop serves inspection purposes (isValid) rather than state-change (invalidate).

 * redesign bookkeeping of validity: rather than propagating invalidness from a container to all children add a method to components to query "inherited invalidness" from their parent containers.  (that probably takes some more thought)

From your user code, however you could consider doing your GUI updates using invokeLater().

**Comment by mabe02** [comment]
Nice one! While we come up with a fix for this, you could do a call to invokeAndWait(..) on the GUI thread object and perform your state mutations there. That will avoid the issue you're seeing now.

**Comment by xazap** [comment]
For the benefit of anyone else running into the same issues as I did, the code below fixed both the issue of the components becoming inaccessible and the deadlock issue:

```java
getTextGUI().getGUIThread().invokeLater(() -> {
	removeAllComponents();
	initComponents(model, locale);
	setFocusedInteractable(contentPane.nextFocus(null));
});
```
Note that the above assumes that the LanternaGUI was started in its own thread:
```java
gui = new MultiWindowTextGUI(new SeparateTextGUIThread.Factory(), screen);
((AsynchronousTextGUIThread) gui.getGUIThread()).start();
```

Thanks for helping out @avl42 and @mabe02!
